README.md
# AWARE regionalized LCIA method for Brightway

This package provides the [AWARE](http://www.wulca-waterlca.org/aware.html) regionalized LCIA method in [brightway](https://brightway.dev>) and [brightway2-regional](https://github.com/brightway-lca/brightway2-regional).

Monthly characterization factors are currently not imported.

## Installation

As this package requires `bw2regional`, we recommend installation via conda.

## Usage

You should be in a project that has the `biosphere3` base biosphere data installed.

```python
import bw2_aware, bw2regional
bw2regional.create_world_collections()
bw2_aware.import_regionalized_aware()
```

The following methods are now available to be used as (either regionalized or site-generic LCIA methods):

* ('AWARE', '1.2 (April 2017)', 'Annual', 'Agricultural')
* ('AWARE', '1.2 (April 2017)', 'Annual', 'Non-agricultural')


CHANGES.md
# Changelog

## 0.3 (2022-04-23)

* Update import functions to follow bw2_lcimpact pattern

## 0.2.1 (2020-06-17)

Bugfix release.

* Change imports for downstream `fiona` changes
* Remove unnecessary imports and code formatting cleanups

## 0.2 (2018-09-17)

Switch sign of water releases, so they count as environmental goods.

## 0.1 (2018-09-12)

First release.


bw2_aware/base.py
from bw2data import Database, Method
from bw2regional import geocollections
import os
import wrapt

data_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), "data"))


class NoRegionalizedSetup(Exception):
    """First run ``bw2regionalsetup` in this project"""

    pass


@wrapt.decorator
def regionalized_setup(wrapped, instance, args, kwargs):
    if "world" not in geocollections:
        raise NoRegionalizedSetup
    return wrapped(*args, **kwargs)


class LCIA:
    def __init__(self, biosphere="biosphere3"):
        self.db = Database(biosphere)
        self.method = Method(self.name)

    @property
    def metadata(self):
        obj = {
            "unit": self.unit,
            "description": self.description,
            "url": self.url,
            "geocollections": [],
        }
        if self.geocollection:
            obj["geocollections"] = [self.geocollection]
        return obj

    @regionalized_setup
    def import_regional_method(self):
        self.method.register(**self.metadata)
        self.setup_geocollections()
        self.method.write(list(self.regional_cfs()))

    def __repr__(self):
        return str(self.name)


bw2_aware/__init__.py
__all__ = (
    "import_global_aware",
    "import_regionalized_aware",
    "create_regionalized_intersections",
)


__version__ = (0, 3)

from .aware import AnnualAgricultural, AnnualNonagricultural
from bw2regional.pandarus_remote import remote, AlreadyExists


METHODS = (AnnualAgricultural, AnnualNonagricultural)



def import_global_aware(biosphere="biosphere3"):
    raise NotImplementedError("This package only provide regionalized CFs")


def import_regionalized_aware(biosphere="biosphere3"):
    for method in METHODS:
        method(biosphere).import_regional_method()


def create_regionalized_intersections():
    try:
        job = remote.calculate_intersection("world", "watersheds-aware")
        job.poll(interval=2)
        if job.status != "finished":
            raise ValueError(
                "Calculation job finished with status '{}'".format(job.status)
            )
    except AlreadyExists:
        pass
    remote.intersection_as_new_geocollection(
        "world", "watersheds-aware", "world-topo-watersheds-aware"
    )
    try:
        job = remote.calculate_intersection(
            "world-topo-watersheds-aware", "watersheds-aware"
        )
        job.poll(interval=2)
        if job.status != "finished":
            raise ValueError(
                "Calculation job finished with status '{}'".format(job.status)
            )
    except AlreadyExists:
        pass
    remote.intersection("world-topo-watersheds-aware", "watersheds-aware")


bw2_aware/aware.py
from .base import LCIA, data_dir, geocollections
from numbers import Number
import os
import fiona


WATER_RESOURCE = [
    ("Water, lake", ("natural resource", "in water")),
    ("Water, river", ("natural resource", "in water")),
    (
        "Water, turbine use, unspecified natural origin",
        ("natural resource", "in water"),
    ),
    ("Water, unspecified natural origin", ("natural resource", "in water")),
    ("Water, cooling, unspecified natural origin", ("natural resource", "in water")),
    ("Water, unspecified natural origin", ("natural resource", "in ground")),
    ("Water, well, in ground", ("natural resource", "in water")),
]

#     Ignored on the advice of Stephan Pfister (ETHZ)
#     ('Water', ('water', 'ground-, long-term')),

WATER_RELEASE = [
    ("Fresh water (obsolete)", ("water", "surface water")),
    ("Water", ("water",)),
    ("Water", ("water", "surface water")),
    ("Water", ("water", "ground-")),
]


class AWARE(LCIA):
    vector_ds = os.path.join(data_dir, "aware.gpkg")
    geocollection = "watersheds-aware"
    unit = "Availability Minus Demand (AMD), m**3/m**2/month"
    description = """AWARE is a water use midpoint indicator representing the relative Available WAter REmaining per area in a watershed, after the demand of humans and aquatic ecosystems has been met. It assesses the potential of water deprivation, to either humans or ecosystems, building on the assumption that the less water remaining available per area, the more likely another user will be deprived.

    It is first calculated as the water Availability Minus the Demand (AMD) of humans and aquatic ecosystems and is relative to the area (m3 m-2 month-1). In a second step, the value is normalized with the world average result (AMD = 0.0136m3m-2 month-1) and inverted, and hence represents the relative value in comparison with the average m3 consumed in the world (the world average is calculated as a consumption-weighted average). Once inverted, 1/AMD  can be interpreted as a surface-time equivalent to generate unused water in this region. The indicator is limited to a range from 0.1 to 100, with a value of 1 corresponding to the world average, and a value of 10, for example, representing a region where there is 10 times less available water remaining per area than the world average."""
    url = "http://www.wulca-waterlca.org/aware.html"
    id_column = "id"

    def _water_flows(self, water):
        for act in self.db:
            if (act["name"], tuple(act["categories"])) in water:
                yield act.key

    def setup_geocollections(self):
        if self.geocollection not in geocollections:
            geocollections[self.geocollection] = {
                "filepath": self.vector_ds,
                "field": self.id_column,
            }

    def global_cfs(self):
        for key in self._water_flows(WATER_RESOURCE):
            yield ((key, self.global_cf, "GLO"))
        for key in self._water_flows(WATER_RELEASE):
            yield ((key, self.global_cf * -1, "GLO"))

    def regional_cfs(self):
        positive_water_flows = list(self._water_flows(WATER_RESOURCE))
        negative_water_flows = list(self._water_flows(WATER_RELEASE))

        for obj in self.global_cfs():
            yield obj

        with fiona.Env():
            with fiona.open(self.vector_ds) as src:
                for feat in src:
                    for key in positive_water_flows:
                        if not isinstance(feat["properties"][self.column], Number):
                            continue
                        yield (
                            key,
                            feat["properties"][self.column],
                            (self.geocollection, feat["properties"][self.id_column]),
                        )

                    for key in negative_water_flows:
                        if not isinstance(feat["properties"][self.column], Number):
                            continue
                        yield (
                            key,
                            feat["properties"][self.column] * -1,
                            (self.geocollection, feat["properties"][self.id_column]),
                        )


class AnnualAgricultural(AWARE):
    column = "annual_agriculture"
    name = ("AWARE", "1.2 (April 2017)", "Annual", "Agricultural")
    global_cf = 45.74


class AnnualNonagricultural(AWARE):
    column = "annual_non_agriculture"
    name = ("AWARE", "1.2 (April 2017)", "Annual", "Non-agricultural")
    global_cf = 20.30
